src/collapse/dtrace.rs

```diff
@@ -109,21 +151,152 @@ impl Collapse for Folder {
 
         None
     }
-}
 
-impl From<Options> for Folder {
-    fn from(opt: Options) -> Self {
-        Self {
-            stack: VecDeque::default(),
-            occurrences: FnvHashMap::with_capacity_and_hasher(512, fnv::FnvBuildHasher::default()),
-            cache_inlines: Vec::new(),
-            opt,
-            stack_str_size: 0,
-        }
+    #[cfg(test)]
+    fn set_nstacks_per_job(&mut self, n: usize) {
+        self.opt.nstacks_per_job = n;
+    }
+
+    #[cfg(test)]
+    fn set_nthreads(&mut self, n: usize) {
+        self.opt.nthreads = n;
     }
 }
 
 impl Folder {
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        assert!(self.occurrences.is_concurrent());
+        assert!(self.opt.nstacks_per_job > 0);
+        assert!(self.opt.nthreads > 1);
+
+        enum Message {
+            Job(Vec<u8>),
+            Shutdown,
+        }
+
+        crossbeam::thread::scope(|scope| {
+            // Spin up the threadpool / worker threads.
+            let (tx_input, rx_input) = channel::unbounded();
```

Comment:
I think this channel should be bounded so we don't send an unbounded amount of work to the workers. It's fine to pace ourselves a little here. Maybe a bound of `2 * self.opt.nthreads`?

Reply:
That's a good idea, but ... 

The following  is more a question for you to see if I understand how channels work. We can use whatever channel bound you think is best. 

... all that's on the channel is an `Option<Vec<u8>>`. Isn't the size of that just (something like)...

* byte (or 2 or whatever rust uses) to differentiate between `None` and `Some` = 1 byte
* pointer = 8 bytes
* length = 8 bytes
* capacity = 8 bytes

...meaning even if we had a channel bound of something like 1024, that would still only be 25k of memory per channel ... a tiny amount. 

I'll put in `2 * self.opt.nthreads` for now. Does what I wrote above make sense, though?

Reply:
Also - The output channel has to remain unbounded, as we don't know in advance how many `io::Result<()>`s need to be sent over it. We could spin up a special thread for collecting errors; so that someone would always be pulling results off the output channel. That error channel could then just send one error back to the main thread in the case that any collapse method failed. But adding that kind of complexity doesn't seem like it adds much. That, coupled with the fact that the memory the output channel will eat up is always going to be tiny (unless you're thinking of a use-case where someone continuously streams data for hours and hours), is what lead me to the design choice in the latest commit, i.e. bound the input channel but leave the output channel unbounded.

Reply:
Nevermind - Now that I get your point about not sending every result (from the comment below), I think I can get us a bounded output channel as well.

Reply:
Yes, you are entirely right that the stuff stored in the channel itself is small, but those `Vec<u8>`s are still taking up space on the heap, and with the current design the relative speed of the parser and the collapsers determines how much heap space we use at any one time. If we bound it, we should see no real downside as far as performance goes, but we _will_significantly reduce memory pressure.

Reply:
Ah - Totally makes sense.